### PR TITLE
chore(scanner): clean resources when done or upon error

### PR DESCRIPTION
**Pull request by RTann**
## Description

Just a cleanup PR.

## Checklist
- [x] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

### Here I tell how I validated my change

no

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: scanner/indexer/indexer.go [hunk 1]
```diff
@@ -76,54 +77,79 @@ type Indexer interface {
...
 // NewIndexer creates a new indexer.
 func NewIndexer(ctx context.Context, cfg config.IndexerConfig) (Indexer, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "scanner/backend/indexer.NewIndexer")
+
 	pool, err := postgres.Connect(ctx, cfg.Database.ConnString, "libindex")
 	if err != nil {
 		return nil, fmt.Errorf("connecting to postgres for indexer: %w", err)
 	}
+	defer func() {
+		if err != nil {
```
Comment (jvdm) [inline scanner/indexer/indexer.go:94]: I've seen a pattern in ClairCore where they explicitly set a variable (e.g., `success := false`) for this. I believe it helps avoid confusion on the different scopes where `err` might be defined and ensures the clean-up is explicit. Wanna do that here, too?

File: scanner/indexer/indexer.go [hunk 2]
```diff
@@ -76,54 +77,79 @@ type Indexer interface {
...
 	if err != nil {
 		return nil, fmt.Errorf("initializing postgres indexer store: %w", err)
 	}
+	defer utils.IgnoreError(func() error {
+		if err != nil {
+			return store.Close(ctx)
+		}
+		return nil
+	})
+
```
Comment (jvdm) [inline scanner/indexer/indexer.go:109]: I've come to appreciate the direct style as in:

```suggestion
	defer func() {
		if err != nil {
			_ = store.Close(ctx)
		}
	}()
```

I agree we are mixing it up now, so fine to leave it. But given this is a clean-up PR, would you like to move out from `utils.IgnoreError()`?

File: scanner/indexer/indexer.go [hunk 3]
```diff
@@ -76,54 +77,79 @@ type Indexer interface {
...
+	ctx = zlog.ContextWithValues(ctx, "component", "scanner/backend/indexer.NewIndexer")
+
 	pool, err := postgres.Connect(ctx, cfg.Database.ConnString, "libindex")
 	if err != nil {
 		return nil, fmt.Errorf("connecting to postgres for indexer: %w", err)
 	}
+	defer func() {
+		if err != nil {
+			pool.Close()
+		}
```
Comment (jvdm) [inline scanner/indexer/indexer.go:96]: I've seen a pattern in ClairCore where they explicitly set a variable (e.g., `success := false`) for this. I believe it helps avoid confusion on the different scopes where `err` might be defined and ensures the clean-up is explicit. Wanna do that here, too?

Comment (RTann) [reply scanner/indexer/indexer.go:96]: sure

File: scanner/indexer/indexer.go (hunk 2 above)
Comment (RTann) [reply scanner/indexer/indexer.go:109]: 👍

File: scanner/indexer/indexer.go (hunk 3 above)
Comment (RTann) [reply scanner/indexer/indexer.go]: https://github.com/stackrox/stackrox/pull/9440/commits/7f9b0e86d81507de277458c30d8c82cb2f89da0a

**Review by jvdm — APPROVED** [review]
LGTM.

**Pull request merged**